Support remaining *_like factory functions for NJT

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144889

Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

jbschlosser committed:
Support remaining *_like factory functions for NJT

[ghstack-poisoned]

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -6109,17 +6109,73 @@ def test_like_shape(self, func):
+                        self.assertEqual(nt.dtype, nt_like.dtype)
+                        self.assertEqual(nt.layout, nt_like.layout)
+                        self.assertEqual(nt.layout, torch.jagged)
+                        self.assertEqual(nt.shape, nt_like.shape)
```
nit: we could factor out the ones that are the same before the if/else

soulitzer reviewed (approved):
Cool!

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i

jbschlosser committed:
Update on "Support remaining *_like factory functions for NJT"


Fixes #144761

This PR adds NJT impls for those *_like functions that were previously missing:
* `full_like()`
* `rand_like()`
* `randint_like()`

It also fixes a bug in existing *_like functions when a new device is specified. Fix is to also transfer `offsets` / `lengths` to the new device.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
